File: Week01/rotate.php
<?php

function rotate(&$nums, $k) {
    if (count($nums) != 1 && $k != 0 && count($nums) != $k){
        $k = $k%count($nums);
        $nums = array_merge(array_slice($nums, 0 - $k, count($nums) - 1), array_slice($nums, 0, count($nums) - $k));
    }
}


function rotate2(&$nums, $k) {
    $count = count($nums);
    $k = $k % $count;
    $this->reverse($nums, 0, $count-1);
    $this->reverse($nums, $k, $count-1);
    $this->reverse($nums, 0, $k-1);
}

function reverse(&$nums, $s, $e){
    while ($s<$e) {
        $tmp = $nums[$s];
        $nums[$s] = $nums[$e];
        $nums[$e] = $tmp;
        $s++;
        $e--;
    }
}

File: Week01/rotate.go
package main

import "fmt"

func main() {
	nums := []int{1,2,3,4,5,6,7}
	rotate(nums,3)
}
func rotate(nums []int, k int) {
	k = k % len(nums)
	reverse(nums[:len(nums)-k])
	fmt.Println(nums)
	reverse(nums[len(nums)-k:])
	fmt.Println(nums)

	reverse(nums)
	fmt.Println(nums)

}

func reverse(nums []int) {
	i, j := 0, len(nums)-1
	for i < j {
		nums[i], nums[j] = nums[j], nums[i]
		i++
		j--

	}
}
